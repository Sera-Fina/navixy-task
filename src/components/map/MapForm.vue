<template>
  <v-form
    ref="markerForm"
    class="map-form px-4"
    lazy-validation
  >
    <v-text-field
      :counter="20"
      :label="$t('form.name')"
      :placeholder="placeholders.name"
      :rules="rules.name"
      :value="form.name"
      required
      @input="onFormChanged('name', $event)"
    ></v-text-field>

    <v-text-field
      :label="$t('form.coordinates')"
      :placeholder="placeholders.coordinates"
      :value="getCoordinatesString(form.coordinates)"
      disabled
      required
    ></v-text-field>
    <v-sheet class="d-flex justify-end flex-wrap">
      <v-btn
        :disabled="loading"
        class="mt-1 success"
        small
        @click="submit"
      >
        {{ $t('buttons.submit') }}
      </v-btn>
    </v-sheet>
  </v-form>
</template>

<script>
import { mapGetters } from 'vuex';
import i18n from '../../plugins/i18n';

export default {
  name: 'MapForm',
  computed: {
    ...mapGetters({
      loading: 'map/markers/getLoading',
      form: 'map/markers/getEditingForm',
    }),
    rules() {
      return {
        name: [
          (value) => !!value || i18n.t('form.required', [i18n.t('form.name')
            .toLowerCase()]),
          (value) => (value && value.length <= 20) || i18n.t('form.maxLength', [20]),
        ],
      };
    },
    placeholders() {
      return {
        name: i18n.t('form.enter', [i18n.t('form.name')
          .toLowerCase()]),
        coordinates: i18n.t('form.enter', [i18n.t('form.coordinates')
          .toLowerCase()]),
      };
    },
  },
  beforeDestroy() {
    this.$store.dispatch('map/markers/setEditingFormFields', {
      name: '',
      coordinates: '',
    });
  },
  methods: {
    onFormChanged(field, value) {
      const form = { [field]: value };
      this.$store.dispatch('map/markers/setEditingFormFields', form);
    },

    getCoordinatesString(coordinates) {
      return coordinates ? `${coordinates.lat}, ${coordinates.lng}` : '';
    },

    async submit() {
      const isValid = this.$refs.markerForm.validate();
      if (isValid) {
        await this.$store.dispatch('map/markers/addMarker', {
          ...this.form,
          coordinatesString: this.getCoordinatesString(this.form.coordinates),
        });
        this.$store.dispatch('map/markers/setEditing', false);
        await this.$store.dispatch('map/markers/uploadMarkers');
      }
    },
  },
};
</script>
