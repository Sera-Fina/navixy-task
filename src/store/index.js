import Vue from 'vue';
import Vuex from 'vuex';
import map from './modules/map';
import locale from './modules/locale';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    locale,
    map,
  },
});
