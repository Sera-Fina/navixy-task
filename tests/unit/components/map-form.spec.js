import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import vuetify from '../../../src/plugins/vuetify';
import MapForm from '../../../src/components/map/MapForm.vue';
import store from '../mocks/store';

describe('MapView.vue', () => {
  let wrapper;

  const localVue = createLocalVue();
  localVue.use(Vuex);

  beforeEach(() => {
    const createdStore = new Vuex.Store({ ...store });
    wrapper = mount(MapForm, {
      localVue,
      vuetify,
      store: createdStore,
      mocks: {
        $t: () => {},
      },
    });
  });

  afterEach(() => {
    wrapper.vm.$destroy();
  });

  it('Should be mounted', (done) => {
    localVue.config.errorHandler = done;
    expect(wrapper)
      .toBeTruthy();
    done();
  });

  it('Should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Should change value on onFormChanged method', () => {
    const storeDispatch = jest.spyOn(wrapper.vm.$store, 'dispatch');
    wrapper.vm.onFormChanged('name', 'New name');
    expect(storeDispatch)
      .toBeCalledWith('map/markers/setEditingFormFields',
        { name: 'New name' });
  });

  it('Should compose right value in getCoordinatesString method', () => {
    const coordinates = wrapper.vm.getCoordinatesString({
      lat: '1',
      lng: '1',
    });
    expect(coordinates)
      .toBe('1, 1');
  });

  it('Should not submits invalid a form', async () => {
    const spy = jest.spyOn(wrapper.vm.$store, 'dispatch');
    await wrapper.vm.submit();
    expect(wrapper.vm.$refs.markerForm.validate()).toBeFalsy();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should submits valid a form', async () => {
    const dispatchSpy = jest.spyOn(wrapper.vm.$store, 'dispatch');
    wrapper.vm.$refs.markerForm.validate = jest.fn().mockResolvedValue(true);
    await wrapper.vm.submit();
    expect(dispatchSpy).toBeCalled();
  });

  it('Should clear store on before destroy', () => {
    const dispatchSpy = jest.spyOn(wrapper.vm.$store, 'dispatch');
    wrapper.vm.$destroy();
    expect(dispatchSpy).toHaveBeenCalledWith('map/markers/setEditingFormFields', {
      name: '',
      coordinates: '',
    });
  });
});
