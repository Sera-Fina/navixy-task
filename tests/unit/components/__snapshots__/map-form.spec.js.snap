// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView.vue Should match snapshot 1`] = `
<form
  class="v-form map-form px-4"
  novalidate="novalidate"
>
  <div
    class="v-input theme--light v-text-field v-text-field--placeholder"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-text-field__slot"
        >
          <input
            id="input-21"
            placeholder="Enter name"
            required="required"
            type="text"
          />
        </div>
      </div>
      <div
        class="v-text-field__details"
      >
        <div
          class="v-messages theme--light"
        >
          <transition-group-stub
            class="v-messages__wrapper"
            name="message-transition"
            tag="div"
          />
        </div>
        <div
          class="v-counter theme--light"
        >
          0 / 20
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="v-input v-input--is-disabled theme--light v-text-field v-text-field--placeholder"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-text-field__slot"
        >
          <input
            disabled="disabled"
            id="input-24"
            placeholder="Enter сoordinates"
            required="required"
            type="text"
          />
        </div>
      </div>
      <div
        class="v-text-field__details"
      >
        <div
          class="v-messages theme--light"
        >
          <transition-group-stub
            class="v-messages__wrapper"
            name="message-transition"
            tag="div"
          />
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="d-flex justify-end flex-wrap v-sheet theme--light"
  >
    <button
      class="mt-1 success v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--small"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        
      
    
      </span>
    </button>
  </div>
</form>
`;
